{% extends "common/base.html" %}
{% block extrastyle %}<style type="text/css">
  .bold-name {font-style: bold; font-weight: bold;}
  .journal {font-style: italic;}
  .comment {font-style: italic;}
  .author {}
  .pub_title {}
</style>
{% endblock %}

{% block title %}Berkeley Phylogenomics Group{% endblock %}

{% block content %}
<h2>Positions in the Sj&ouml;lander lab</h2> 

<p>
We are looking for bright, motivated and talented individuals with various backgrounds and levels
of expertise to join a group of students, postdoctoral scholars and programmers in the
development of an open-source bioinformatics platform for biological discovery. This resource
will build on our existing PhyloFacts database and web servers, and will be deployed to assist
researchers in TB and other human pathogens in identifying drug targets and development of
effective drugs and diagnostics, to interpret human genetic variation, and to contribute to basic
research in evolution and protein structure and function. Our needs are quite broad; in addition
to programmer positions we have openings for research associates and postdoctoral scholars.
</p>

<p>
This position is located on the UC Berkeley campus in the beautiful Stanley Hall at the north-
east corner of campus, in the Institute for Quantitative Biomedical Research (QB3); it is funded
by the Department of Energy Systems Biology Knowledgebase. The Sjölander lab includes
programmers, postdocs and students from a variety of departments across campus. We
collaborate with scientists at different universities and research labs in The Netherlands, France,
Spain, England, Sweden, India, Japan, and Brazil. Members of this group have opportunities
to participate in a collaboration with Google in the development of an open-source
system for bioinformatics and biological discovery.
</p>

<br />
<h3>Paid student internships at Google</h3> 
The Sjölander lab and Google are collaborating in the development of an open-source bioinformatics platform integrating bioinformatics methods and data associated with gene families across the Tree of Life. Paid internships are available immediately for students with appropriate backgrounds, and additional internships will be available throughout 2012. Students with various skills and backgrounds are needed, including databases, software engineering, statistical modeling and graphical user interface design. Required background/skills include: Python programming, coursework on algorithms and data structures, and excellent communication skills. Training in bioinformatics will be provided.
<br /> 
<p>Students who are interested in participating in this exciting project should do two things: 
<br />
1) Let Prof. Sjolander know that you are applying (email kimmen 'at' berkeley 'dot' edu)<br />
2) <a href="http://www.google.com/intl/en/jobs/students/tech/internships/uscanada/">Apply via the Google website</a>, letting Google know that you are interested in the joint project with UC Berkeley Prof. Sjolander.
</p>
<br />
<br />

<h3>Applications Programmers</h3> 
We need individuals with experience in one or more of the following areas: webserver development, graphical user interfaces, Postgres/MySQL database development, distributed processes, cloud computing, systems administration and scientific programming. We use agile programming methods, including pair programming. No prior experience in bioinformatics is required; on-the-job training in the underlying bioinformatics methods, theory and data structures will be provided. To apply for this position, go <a href="https://hrw-vip-prod.is.berkeley.edu/psp/JOBSPROD/EMPLOYEE/HRMS/c/HRS_HRAM.HRS_CE.GBL?Page=HRS_CE_JOB_DTL&Action=A&JobOpeningId=12923&SiteId=1&PostingSeq=1">here</a>.
<br />
<a id="google-berkeley-more" href="#" onclick = 
								"document.getElementById('google-berkeley').style.display = 'block'; this.style.display = 'none'; return false;">More details ...</a>
<div id="google-berkeley" style="display:none">
<br />
<b>Primary Responsibilities:</b>
<ul type="disc">
<li>Writing software to populate our bioinformatics database. Software will include short scripts to download data from external sources, and to run third-party software tools to construct hidden Markov models (HMMs), phylogenetic trees for gene families, multiple sequence alignments, and other data. Some software development will involve more complex scientific programming, which will be performed jointly (in pair-programming) with postdoctoral scholars and graduate students in the Sjolander lab.</li>
<li>Modifications to webpages to improve accessibility and interpretability of data for external users, including the development and installation of intuitive graphical user interfaces and navigation tools.</li>
<li>MySQL or Postgres database development and administration, including schema modifications and database query construction and optimization.
<li>Assisting in the release of our software tools in open source.</li>
</ul>
<br />
<b>Required Qualifications</b>
<ul type="disc">
<li>Minimum 2 years work experience as an applications programmer (preferably in industry).</li>
<li>Bachelor's degree in science or engineering, or related training.</li>
<li>Experience with developing web applications, web programming languages and object oriented programming concepts.</li>
<li>Strong Python programming skills (Django experience is a plus).</li>
<li>Expertise with relational databases (MySQL or Postgres).</li>
<li>Strong Linux/Unix skills.</li>
<li>Strong written/oral communication skills.</li>
<li>Ability to juggle multiple tasks simultaneously.</li>
<li>Ability to develop and implement quality assurance/quality control.</li>
<li>Ability to work with software developed by others and to debug this code as needed.</li>
<li>Ability to write clear, modular and well-documented code.</li>
<li>Must be self motivated, able to learn quickly, meet deadlines and demonstrate problem solving skills.</li>
<li>Social skills: we work closely together, and the ability to get along with people from different backgrounds is essential to project success.</li>
</ul>
</br>
<a href="#" onclick = "document.getElementById('google-berkeley').style.display = 'none'; 
					document.getElementById('google-berkeley-more').style.display = 'inline';
					return false;">less details ...</a>
<br />
</div>
<br />
<br />
<h3>Postdoctoral Scholar positions</h3> 
Applicants for postdoctoral scholar positions must have a doctoral degree in biology, bioinformatics or computer science (or related fields). Python programming skills are required, but need not be at the same level as the programmer positions. Bioinformatics experience is a plus but is not required. Applicants for the postdoctoral scholar positions should include in their application a published paper for which they were the first author, describing their role on the paper.
<br />
<br />
<p>
To apply, please send your <b>resume</b> and <b>cover letter</b> to to Dr. Ajithkumar Warrier ("awarrier" at "berkeley.edu"). Your cover letter should describe the job you want to apply for (whether programmer or postdoctoral scholar), your career objectives, what you consider your greatest strengths, and those aspects of the job that appeal to you most. We have many different ways to use talented programmers and postdoctoral scholars, and can try to tune a position to the applicant skills and interests.
</p>

<br />

{% endblock %}
