{% extends "common/base.html" %}
{% block content %} 
<script type="text/javascript" src="/static/js/phylofacts.js"></script>

<h2>UniProt Sequences</h2>
<ul>
  <li>Homologs of <a href = "homologs_pqqA.txt">PqqA</a>, <a href = "homologs_pqqB.txt">PqqB</a>, <a href = "homologs_pqqC.txt">PqqC</a>, <a href = "homologs_pqqD.txt">PqqD</a>, <a href = "homologs_pqqE.txt">PqqE</a>, <a href = "homologs_pqqF.txt">PqqF</a></li>
</ul>

<h2>Multiple Sequence Alignments</h2>

<p>
<i>Note: Multiple sequence alignments are shown in UCSC Aligned Fasta format. The alignments can be retrieved using the hyperlinks, or viewed using the JalView JAVA-based alignment viewers buttons at right. 
</i>
</p>
<ul>
<li>PqqA <a href = "MSA_pqqA.afa">MSA</a>
	<applet archive='/static/java/jalviewApplet.jar' name="Jalview"
		code='jalview.bin.JalviewLite' 
		width='90' 
		height='35'
		>
		<param name='file' 
			value='/supplementary/pqq/MSA_pqqA.afa' />
		<param name='label' value='Jalview' />
		<param name='userDefinedColourr' value='C=yellow; R,K,H=FF5555; D,E=5555FF' />
		<param name='showFullId' value='false' />
		<param name='showTreeDistances' value='false' />
		<param name='showTreeBootstraps' value='false' />
		<param name='linkLabel_1' value='SRS' />
		<param name='linkUrl_1' value='http://srs.ebi.ac.uk/srs7bin/cgi-bin/wgetz?-e+[uniprot-all:$SEQUENCE_ID$]+-vn+2' />
		<param name='linkLabel_2' value='Uniprot' />
		<param name='linkUrl_2' value='http://us.expasy.org/cgi-bin/niceprot.pl?$SEQUENCE_ID$' />
		<param name='APPLICATION_URL' value='http://www.jalview.org/services/launchApp' />If the <a href='http://www.jalview.org/' target='_blank'>Jalview</a> applet failed to load; please <a href='http://java.com/en/download/'>download Java</a>.
	</applet>
</li> 
<li>PqqB <a href = "MSA_pqqB.afa">MSA</a>
	<applet archive='/static/java/jalviewApplet.jar' name="Jalview"
		code='jalview.bin.JalviewLite' 
		width='90' 
		height='35'
		>
		<param name='file' 
			value='/supplementary/pqq/MSA_pqqB.afa' />
		<param name='label' value='Jalview' />
		<param name='userDefinedColourr' value='C=yellow; R,K,H=FF5555; D,E=5555FF' />
		<param name='showFullId' value='false' />
		<param name='showTreeDistances' value='false' />
		<param name='showTreeBootstraps' value='false' />
		<param name='linkLabel_1' value='SRS' />
		<param name='linkUrl_1' value='http://srs.ebi.ac.uk/srs7bin/cgi-bin/wgetz?-e+[uniprot-all:$SEQUENCE_ID$]+-vn+2' />
		<param name='linkLabel_2' value='Uniprot' />
		<param name='linkUrl_2' value='http://us.expasy.org/cgi-bin/niceprot.pl?$SEQUENCE_ID$' />
		<param name='APPLICATION_URL' value='http://www.jalview.org/services/launchApp' />If the <a href='http://www.jalview.org/' target='_blank'>Jalview</a> applet failed to load; please <a href='http://java.com/en/download/'>download Java</a>.
	</applet>

</li>
<li>PqqC <a href = "MSA_pqqC.afa">MSA</a>
	<applet archive='/static/java/jalviewApplet.jar' name="Jalview"
		code='jalview.bin.JalviewLite' 
		width='90' 
		height='35'
		>
		<param name='file' 
			value='/supplementary/pqq/MSA_pqqC.afa' />
		<param name='label' value='Jalview' />
		<param name='userDefinedColourr' value='C=yellow; R,K,H=FF5555; D,E=5555FF' />
		<param name='showFullId' value='false' />
		<param name='showTreeDistances' value='false' />
		<param name='showTreeBootstraps' value='false' />
		<param name='linkLabel_1' value='SRS' />
		<param name='linkUrl_1' value='http://srs.ebi.ac.uk/srs7bin/cgi-bin/wgetz?-e+[uniprot-all:$SEQUENCE_ID$]+-vn+2' />
		<param name='linkLabel_2' value='Uniprot' />
		<param name='linkUrl_2' value='http://us.expasy.org/cgi-bin/niceprot.pl?$SEQUENCE_ID$' />
		<param name='APPLICATION_URL' value='http://www.jalview.org/services/launchApp' />If the <a href='http://www.jalview.org/' target='_blank'>Jalview</a> applet failed to load; please <a href='http://java.com/en/download/'>download Java</a>.
	</applet>

</li>
<li>PqqD <a href = "MSA_pqqD.afa">MSA</a>
	<applet archive='/static/java/jalviewApplet.jar' name="Jalview"
		code='jalview.bin.JalviewLite' 
		width='90' 
		height='35'
		>
		<param name='file' 
			value='/supplementary/pqq/MSA_pqqD.afa' />
		<param name='label' value='Jalview' />
		<param name='userDefinedColourr' value='C=yellow; R,K,H=FF5555; D,E=5555FF' />
		<param name='showFullId' value='false' />
		<param name='showTreeDistances' value='false' />
		<param name='showTreeBootstraps' value='false' />
		<param name='linkLabel_1' value='SRS' />
		<param name='linkUrl_1' value='http://srs.ebi.ac.uk/srs7bin/cgi-bin/wgetz?-e+[uniprot-all:$SEQUENCE_ID$]+-vn+2' />
		<param name='linkLabel_2' value='Uniprot' />
		<param name='linkUrl_2' value='http://us.expasy.org/cgi-bin/niceprot.pl?$SEQUENCE_ID$' />
		<param name='APPLICATION_URL' value='http://www.jalview.org/services/launchApp' />If the <a href='http://www.jalview.org/' target='_blank'>Jalview</a> applet failed to load; please <a href='http://java.com/en/download/'>download Java</a>.
	</applet>
</li>
<li>PqqE <a href = "MSA_pqqE.afa">MSA</a>
	<applet archive='/static/java/jalviewApplet.jar' name="Jalview"
		code='jalview.bin.JalviewLite' 
		width='90' 
		height='35'
		>
		<param name='file' 
			value='/supplementary/pqq/MSA_pqqE.afa' />
		<param name='label' value='Jalview' />
		<param name='userDefinedColourr' value='C=yellow; R,K,H=FF5555; D,E=5555FF' />
		<param name='showFullId' value='false' />
		<param name='showTreeDistances' value='false' />
		<param name='showTreeBootstraps' value='false' />
		<param name='linkLabel_1' value='SRS' />
		<param name='linkUrl_1' value='http://srs.ebi.ac.uk/srs7bin/cgi-bin/wgetz?-e+[uniprot-all:$SEQUENCE_ID$]+-vn+2' />
		<param name='linkLabel_2' value='Uniprot' />
		<param name='linkUrl_2' value='http://us.expasy.org/cgi-bin/niceprot.pl?$SEQUENCE_ID$' />
		<param name='APPLICATION_URL' value='http://www.jalview.org/services/launchApp' />If the <a href='http://www.jalview.org/' target='_blank'>Jalview</a> applet failed to load; please <a href='http://java.com/en/download/'>download Java</a>.
	</applet>
</li>
<li>PqqF <a href = "MSA_pqqF.afa">MSA</a>
	<applet archive='/static/java/jalviewApplet.jar' name="Jalview"
		code='jalview.bin.JalviewLite' 
		width='90' 
		height='35'
		>
		<param name='file' 
			value='/supplementary/pqq/MSA_pqqF.afa' />
		<param name='label' value='Jalview' />
		<param name='userDefinedColourr' value='C=yellow; R,K,H=FF5555; D,E=5555FF' />
		<param name='showFullId' value='false' />
		<param name='showTreeDistances' value='false' />
		<param name='showTreeBootstraps' value='false' />
		<param name='linkLabel_1' value='SRS' />
		<param name='linkUrl_1' value='http://srs.ebi.ac.uk/srs7bin/cgi-bin/wgetz?-e+[uniprot-all:$SEQUENCE_ID$]+-vn+2' />
		<param name='linkLabel_2' value='Uniprot' />
		<param name='linkUrl_2' value='http://us.expasy.org/cgi-bin/niceprot.pl?$SEQUENCE_ID$' />
		<param name='APPLICATION_URL' value='http://www.jalview.org/services/launchApp' />If the <a href='http://www.jalview.org/' target='_blank'>Jalview</a> applet failed to load; please <a href='http://java.com/en/download/'>download Java</a>.
	</applet>
</li>
<li></li>
<li>Pairwise alignment used to construct the homology models of PqqB, PqqC, PqqD, PqqE,</li>
</ul>

<h2>Homology Models</h2>
<b>PDB-formatted structure files.</b> These can be saved onto your local machine and examined using Rasmol, Chimera, or other structure visualization tools. 
<ul>
  <li><a href = "models_pqqB.pdb">PqqB</a>, <a href ="models_pqqC.pdb">PqqC</a>, <a href = "models_pqqD.pdb">PqqD</a>, <a href ="models_pqqE.pdb">PqqE</a></li>
</ul>

<h2>Phylofacts</h2>
<ul>
  <li>PhyloFact books for PqqA, PqqB, PqqC, PqqD, PqqE, PqqF </li>
</ul>

<h2>Figures</h2>
<ul>
<li><a href="#figure1">Figure 1</a>Venn diagram of core Pqq biosynthesis genes (PqqC, PqqD and PqqE)</li>
<li><a href="#figure2">Figure 2</a>Multiple sequence alignment of predicted PqqB from Bacillus pseudomycoides (top of the alignment) and species that known to synthesize PQQ</li>
<li>Figure 3 Structure of PqqB (from Pseudomonas putida) with conserved residues</li>
<li>Figure 4 Conserved functional residues (blue) mapped to the PqqC structure (1otw)</li>
<li>Figure 5 Pqq C structure (1otw, from Klebsiella pneumoniae) with the well conserved residues very close to the PQQ (less than 5-6 A). Green: residues shown the biochemical importance; pink: residues conserved in all seed sequences 
</li>
</ul>

<h3>Figure 1 Venn diagram of core Pqq biosynthesis genes (PqqC, PqqD and PqqE)</h3>
<p>
  <img name="figure1" src="/static/apps/supplementary/img/pqq/figure1.jpg" />
  </a>
</p>

<h3>Figure 2 Multiple sequence alignment of predicted PqqB from Bacillus pseudomycoides (top of the alignment) and species that known to synthesize PQQ</h3>
<p>
  <img name="figure2" src="/static/apps/supplementary/img/pqq/figure2.jpg" />
  </a>
</p>

<h3>Figure 3. Structure of PqqB (from Pseudomonas putida) with conserved residues</h3>

<h3>Figure 4.</h3> 

<h3>Figure 5. </h3>

<h3>Figure 6. </h3>

{% endblock %}
