{% extends "common/legacy_base.html" %}

{% block mid_menu %}{% include "satchmo/mid_menu.html" %}{% endblock %}

{% block app_logo %}<a href='/q/satchmo/'><span>SATCHMO-JS</span></a>{% endblock %}
{% block statcounter %}{% include "satchmo/statcounter.html" %}{% endblock %}

{% block title %}{% include "satchmo/title.html" %}{% endblock %}

{% block body_title %}SATCHMO-JS Supplementary Material{% endblock %}

{% block extrascripts %}{% endblock %}

{% block extrastyle %}
<style type='text/css'>
#content {
    margin: 0 1em;
}
p, caption {
    line-height: 105%;
}
</style>
{% endblock %}

{# block toplinks #}
{# include "snippets/menu.html" #}
{# endblock #}

{% block content %}
<div>
    <h3 class="ktitle">Supplementary Material for</h3>
    <h1 class="ktitle">SATCHMO-JS: A webserver for simultaneous protein
        multiple sequence alignment and tree construction
    </h1>
    <p class='ktitle'>Raffi Hagopian, John R. Davidson, Ruchira S. Datta, Bushra Samad, Glen R. Jarvis and Kimmen Sj&ouml;lander<br /><a href='http://nar.oxfordjournals.org/content/vol38/suppl_2/index.dtl'>Nucleic Acids Research Web Server Issue, 2010</a></p>
    <h3 class='ktitle'>Please visit the <a href='/q/satchmo/'>SATCHMO-JS webserver</a>.</h3>
</div>

<h2>Table of Contents</h2>
<ol id='toc'>
  <li><a href='#id_intro'>Introduction</a></li>
  <li><a href='#id_eval'>Evaluation Scores</a>
      <ol>
          <li><a href='#id_qdev'><span title='TP / (TP+FN)'>Q_Developer</span></a></li>
          <li><a href='#id_qmod'><span title='TP / (TP+FP)'>Q_Modeler</span></a></li>
          <li><a href='#id_clineshift'>Cline Shift</a></li>
          <li><a href='#id_qcom'>Q_Combined</a></li>
      </ol>
  </li>
  <li><a href='#id_downloads'>Datasets Used in Validation Experiments And Other Downloads</a></li>
</ol>

<h2 id='id_intro'>Introduction</h2>
<p>
Results on a benchmark dataset of 983 structurally
aligned pairs from the PREFAB benchmark dataset [<a href='#ref1'>1</a>]
show that SATCHMO-JS provides a statistically significant improvement in
alignment accuracy over MUSCLE, MAFFT, ClustalW and the original SATCHMO
algorithm (see table 1).
</p>

<p>
Additional figures and data are provided below.
</p>

<br />
<table class='datatable'>
    <caption><b>Table 1: SATCHMO-JS Performance Relative to Other Methods.</b> {% include 'satchmo/table_text.html'%}</caption>
    <tbody>
        <tr>
            <th></th>
            <th>SATCHMO</th>
            <th>MAFFT</th>
            <th>MUSCLE</th>
            <th>CLUSTALW</th>
        </tr>
        <tr>
            <th><span title='TP / (TP+FN)'>Q_Developer</span></th>
            <td>3.52E-13</td>
            <td>1.14E-05</td>
            <td>0.010</td>
            <td>7.16E-57</td>
        </tr>
        <tr>
            <th><span title='TP / (TP+FP)'>Q_Modeler</span></th>
            <td>8.66E-05</td>
            <td>0.204</td>
            <td>8.83E-06</td>
            <td>7.42E-62</td>
        </tr>
        <tr>
            <th>Cline</th>
            <td>1.00E-10</td>
            <td>0.157</td>
            <td>0.007</td>
            <td>1.05E-51</td>
        </tr>
        <tr>
            <th>Q_Combined</th>
            <td>4.30E-12</td>
            <td>0.093</td>
            <td>0.004</td>
            <td>1.76E-56</td>
    </tbody>
</table>
<br />

<p>
The table above shows the p-values for the significance of the improvement
of SATCHMO-JS relative to the original SATCHMO, MUSCLE, MAFFT and ClustalW.
P-values were computed on the dataset as a whole (all 983 reference pairs)
using Wilcoxon Paired Signed Rank Tests.
</p>
<h2 id='id_eval'>Evaluation Score Results</h2>

<h3 id='id_qdev'><span title='TP / (TP+FN)'>Q_Developer</span></h3>

<p>
The <span title='TP / (TP+FN)'>Q_Developer</span> score measures recall. It is described in Wang and Dunbrack [<a href='#ref2'>2</a>].
</p>

<p>
    <span title='TP / (TP+FN)'>Q_Developer</span> = TP / (TP+FN)
    <br />
    TP = # of correctly aligned residue pairs in the sequence
    alignment (i.e., agree with the reference alignment).
    <br />
    FN = # of aligned residue pairs in the reference alignment which are
    not in the sequence alignment (i.e., they are missed by the sequence
    alignment)
</p>

<p>
    In other words, the <span title='TP / (TP+FN)'>Q_Developer</span> score measures the fraction of the
    reference alignment that is correctly predicted by the sequence
    alignment, and is thus a measure of the recall.<br />
</p>

<img src="/static/apps/satchmo/img/supplemental_materials/q_developer_plot.png"
     alt='Q-Developer Plot' class='sup_img' />


<br />
<table class='datatable'>
    <caption><b>Table 2: <span title='TP / (TP+FN)'>Q_Developer</span> Score.</b> {% include 'satchmo/table_text.html' %}</caption>
    <tbody>
        <tr>
            <th>% ID</th>
            <th>CLUSTAL-W</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO-JS</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>0.000</td>
            <td>0.050</td>
            <td>0.012</td>
            <td>0.023</td>
            <td>0.052</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.085</td>
            <td>0.100</td>
            <td>0.099</td>
            <td>0.170</td>
            <td>0.140</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>0.257</td>
            <td>0.325</td>
            <td>0.330</td>
            <td>0.371</td>
            <td>0.378</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>0.361</td>
            <td>0.483</td>
            <td>0.499</td>
            <td>0.565</td>
            <td>0.566</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>0.504</td>
            <td>0.672</td>
            <td>0.640</td>
            <td>0.692</td>
            <td>0.638</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>0.647</td>
            <td>0.771</td>
            <td>0.761</td>
            <td>0.787</td>
            <td>0.734</td>
            </tr>
        <tr>
            <td>30-35</td>
            <td>0.712</td>
            <td>0.834</td>
            <td>0.817</td>
            <td>0.857</td>
            <td>0.784</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.739</td>
            <td>0.853</td>
            <td>0.846</td>
            <td>0.900</td>
            <td>0.843</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.918</td>
            <td>0.927</td>
            <td>0.944</td>
            <td>0.966</td>
            <td>0.923</td>
        </tr>
    </tbody>
</table>
<br />
<table class='datatable'>
    <caption><b>Table 3: P-values For <span title='TP / (TP+FN)'>Q_Developer</span> Score.</b> p-values were computed using Wilcoxon paired signed rank test.</caption>
    <tbody>
        <tr>
            <th>% ID</th>
            <th>CLUSTALW</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>1</td>
            <td>1</td>
            <td>1</td>
            <td>1</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.007334</td>
            <td>0.02253</td>
            <td>0.0267</td>
            <td>0.06337</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>1.337e-07</td>
            <td>0.0636</td>
            <td>0.06515</td>
            <td>0.872</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>3.342e-17</td>
            <td>0.03935</td>
            <td>0.03105</td>
            <td>0.609</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>1.811e-16</td>
            <td>0.9575</td>
            <td>0.02304</td>
            <td>7.863e-06</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>3.65e-11</td>
            <td>0.4613</td>
            <td>0.2609</td>
            <td>5.686e-06</td>
        </tr>
        <tr>
            <td>30-35</td>
            <td>3.356e-05</td>
            <td>0.9755</td>
            <td>0.4267</td>
            <td>8.368e-05</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.0001994</td>
            <td>0.7154</td>
            <td>0.6006</td>
            <td>0.002895</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.0003571</td>
            <td>0.001997</td>
            <td>0.03125</td>
            <td>1.936e-08</td>
        </tr>
    </tbody>
</table>
<br />

<h3 id='id_qmod'><span title='TP / (TP+FP)'>Q_Modeler</span></h3>


<p>
The <span title='TP / (TP+FP)'>Q_Modeler</span> score measures precision. It is described in Wang and Dunbrack
[<a href="#ref2">2</a>].
</p>


<p>
  <span title='TP / (TP+FP)'>Q_Modeler</span> = TP / (TP + FP)
  <br />
  TP = # of correctly aligned residue pairs in the sequence alignment (i.e,
  that agree with the reference) 
  <br />
  FP = # of incorrectly aligned residue pairs in the sequence alignment
  (i.e., pairs in the sequence alignment that are not aligned in the
  reference)
</p>

<img src="/static/apps/satchmo/img/supplemental_materials/q_modeler_plot.png"
     alt='Q-Modeler Plot' class='sup_img' />

<br />
<table class='datatable'>
    <caption><b>Table 4: <span title='TP / (TP+FP)'>Q_Modeler</span> Score.</b> {% include 'satchmo/table_text.html' %}</caption>
    <tbody>
        <tr>
            <th>% ID Bin</th>
            <th>CLUSTAL-W</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO-JS</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>0.000</td>
            <td>0.013</td>
            <td>0.006</td>
            <td>0.006</td>
            <td>0.020</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.055</td>
            <td>0.066</td>
            <td>0.082</td>
            <td>0.116</td>
            <td>0.097</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>0.226</td>
            <td>0.276</td>
            <td>0.312</td>
            <td>0.319</td>
            <td>0.338</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>0.335</td>
            <td>0.444</td>
            <td>0.495</td>
            <td>0.527</td>
            <td>0.543</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>0.497</td>
            <td>0.639</td>
            <td>0.640</td>
            <td>0.658</td>
            <td>0.642</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>0.651</td>
            <td>0.754</td>
            <td>0.772</td>
            <td>0.774</td>
            <td>0.750</td>
        </tr>
        <tr>
            <td>30-35</td>
            <td>0.686</td>
            <td>0.796</td>
            <td>0.801</td>
            <td>0.806</td>
            <td>0.783</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.691</td>
            <td>0.805</td>
            <td>0.803</td>
            <td>0.834</td>
            <td>0.813</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.858</td>
            <td>0.867</td>
            <td>0.885</td>
            <td>0.901</td>
            <td>0.876</td>
        </tr>
    </tbody>
</table>
<br />
<table class='datatable'>
    <caption><b>Table 5: P-values for <span title='TP / (TP+FP)'>Q_Modeler</span> Score.</b> p-values were computed using Wilcoxon paired signed rank test.</caption>
    <tbody>
        <tr>
            <th>% ID Bin</th>
            <th>CLUSTALW</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>1</td>
            <td>1</td>
            <td>1</td>
            <td>1</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.003185</td>
            <td>0.01801</td>
            <td>0.2069</td>
            <td>0.04501</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>6.606e-08</td>
            <td>0.02383</td>
            <td>0.5326</td>
            <td>0.2551</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>3.02e-20</td>
            <td>0.0009527</td>
            <td>0.6601</td>
            <td>0.7283</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>3.609e-17</td>
            <td>0.4847</td>
            <td>0.2169</td>
            <td>0.009542</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>1.529e-10</td>
            <td>0.6399</td>
            <td>0.3065</td>
            <td>0.003368</td>
        </tr>
        <tr>
            <td>30-35</td>
            <td>4.937e-05</td>
            <td>0.7914</td>
            <td>0.332</td>
            <td>0.005784</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.0001482</td>
            <td>0.8566</td>
            <td>0.5382</td>
            <td>0.1042</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.0002237</td>
            <td>9.56e-06</td>
            <td>0.03163</td>
            <td>8.777e-07</td>
        </tr>
    </tbody>
</table>
<br />
<h3 id='id_clineshift'>Cline Shift</h3>

<p>
The Cline Shift score includes a small positive score for being close to
the reference alignment, and a small penalty for overalignment. It is
described in (Cline et al., 2002) [<a href="#ref3">3</a>].
</p>

<img src="/static/apps/satchmo/img/supplemental_materials/cline_shift_score.png"
     alt='Cline Shift Plot' class='sup_img' />

<br />
<table class='datatable'>
    <caption><b>Table 6: Cline Shift Score.</b> {% include 'satchmo/table_text.html' %}</caption>
    <tbody>
        <tr>
            <th>% ID Bin</th>
            <th>CLUSTAL-W</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO-JS</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>-0.016</td>
            <td>0.033</td>
            <td>-0.009</td>
            <td>0.001</td>
            <td>0.001</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.030</td>
            <td>0.035</td>
            <td>0.060</td>
            <td>0.113</td>
            <td>0.074</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>0.235</td>
            <td>0.287</td>
            <td>0.314</td>
            <td>0.332</td>
            <td>0.353</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>0.355</td>
            <td>0.465</td>
            <td>0.502</td>
            <td>0.549</td>
            <td>0.556</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>0.511</td>
            <td>0.659</td>
            <td>0.644</td>
            <td>0.678</td>
            <td>0.644</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>0.673</td>
            <td>0.775</td>
            <td>0.780</td>
            <td>0.795</td>
            <td>0.752</td>
        </tr>
        <tr>
            <td>30-35</td>
            <td>0.712</td>
            <td>0.823</td>
            <td>0.816</td>
            <td>0.840</td>
            <td>0.791</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.727</td>
            <td>0.830</td>
            <td>0.830</td>
            <td>0.878</td>
            <td>0.831</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.887</td>
            <td>0.897</td>
            <td>0.912</td>
            <td>0.930</td>
            <td>0.898</td>
        </tr>
    </tbody>
</table>
<br />
<table class='datatable'>
    <caption><b>Table 7: P-values for Cline Shift Score.</b> P-values were computed using Wilcoxon paired signed rank test.</caption>
    <tbody>
        <tr>
            <th>% ID Bin</th>
            <th>CLUSTALW</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>0.6875</td>
            <td>0.8438</td>
            <td>0.5625</td>
            <td>0.8125</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.005889</td>
            <td>0.02854</td>
            <td>0.3122</td>
            <td>0.06204</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>1.60E-05</td>
            <td>0.174</td>
            <td>0.9787</td>
            <td>0.5443</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>1.96E-16</td>
            <td>0.04465</td>
            <td>0.5732</td>
            <td>0.7483</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>1.28E-15</td>
            <td>0.8633</td>
            <td>0.4243</td>
            <td>9.68E-05</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>1.44E-10</td>
            <td>0.7866</td>
            <td>0.5926</td>
            <td>6.02E-06</td>
        </tr>
        <tr>
            <td>30-35</td>
            <td>2.86E-05</td>
            <td>0.8849</td>
            <td>0.672</td>
            <td>0.0003387</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.0004037</td>
            <td>0.7343</td>
            <td>0.7518</td>
            <td>0.00375</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.0005661 </td>
            <td>0.0006641</td>
            <td>0.03723</td>
            <td>4.45E-09</td>
        </tr>
    </tbody>
</table>
<br />

<h3 id='id_qcom'>Q_Combined</h3>
<p>
The Q_Combined score penalizes over- and under-alignment, and measures both
the precision and recall. It is described in Wang and Dunbrack [<a
href="#ref2">2</a>].
</p>

<img src="/static/apps/satchmo/img/supplemental_materials/q_combined_score.png" 
     alt='Q-Combined Score Plot' class='sup_img' />

<br />
<table class='datatable'>
    <caption><b>Table 8: Q_Combined Score.</b> {% include 'satchmo/table_text.html' %}</caption>
    <tbody>
        <tr>
            <th>% ID Bin</th>
            <th>CLUSTALW</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO-JS</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>0.000</td>
            <td>0.011</td>
            <td>0.004</td>
            <td>0.005</td>
            <td>0.016</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.016</td>
            <td>0.024</td>
            <td>0.023</td>
            <td>0.035</td>
            <td>0.024</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>0.111</td>
            <td>0.154</td>
            <td>0.171</td>
            <td>0.173</td>
            <td>0.164</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>0.209</td>
            <td>0.288</td>
            <td>0.317</td>
            <td>0.326</td>
            <td>0.328</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>0.312</td>
            <td>0.423</td>
            <td>0.417</td>
            <td>0.442</td>
            <td>0.392</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>0.459</td>
            <td>0.569</td>
            <td>0.559</td>
            <td>0.575</td>
            <td>0.529</td>
        </tr>
        <tr>
            <td>30-35</td>
            <td>0.573</td>
            <td>0.686</td>
            <td>0.681</td>
            <td>0.706</td>
            <td>0.634</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.617</td>
            <td>0.734</td>
            <td>0.739</td>
            <td>0.778</td>
            <td>0.718</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.831</td>
            <td>0.836</td>
            <td>0.859</td>
            <td>0.877</td>
            <td>0.828</td>
        </tr>
    </tbody>
</table>
<br />
<table class='datatable'>
    <caption><b>Table 9: P-values for Q_Combined Score.</b>  P-values were computed using Wilcoxon paired signed rank test.</caption>
    <tbody>
        <tr>
            <th>% ID Bin</th>
            <th>CLUSTALW</th>
            <th>MUSCLE</th>
            <th>MAFFT</th>
            <th>SATCHMO</th>
        </tr>
        <tr>
            <td>0-5</td>
            <td>1</td>
            <td>1</td>
            <td>1</td>
            <td>1</td>
        </tr>
        <tr>
            <td>5-10</td>
            <td>0.185</td>
            <td>0.3604</td>
            <td>0.408</td>
            <td>0.3575</td>
        </tr>
        <tr>
            <td>10-15</td>
            <td>5.243e-07</td>
            <td>0.5067</td>
            <td>0.8975</td>
            <td>0.7249</td>
        </tr>
        <tr>
            <td>15-20</td>
            <td>6.631e-17</td>
            <td>0.06909</td>
            <td>0.9918</td>
            <td>0.8895</td>
        </tr>
        <tr>
            <td>20-25</td>
            <td>4.484e-17</td>
            <td>0.3209</td>
            <td>0.2309</td>
            <td>3.426e-05</td>
        </tr>
        <tr>
            <td>25-30</td>
            <td>4.683e-10</td>
            <td>0.9803</td>
            <td>0.3625</td>
            <td>0.0004511</td>
        </tr>
        <tr>
            <td>30-35</td>
            <td>5.037e-05</td>
            <td>0.626</td>
            <td>0.5964</td>
            <td>3.792e-05</td>
        </tr>
        <tr>
            <td>35-40</td>
            <td>0.0001537</td>
            <td>0.8443</td>
            <td>0.8508</td>
            <td>0.004093</td>
        </tr>
        <tr>
            <td>40-100</td>
            <td>0.0002401</td>
            <td>0.0003652</td>
            <td>0.02154</td>
            <td>2.856e-09</td>
        </tr>
    </tbody>
</table>
<br />

<!--<p>To Appear in NAR Web Server Issue 2010:</p>
<ul class='refs'>
    <li>Hagopian, R., Davidson, J., Datta, R., Samad, B., Jarvis, G., and <b>Sj&ouml;lander, K.</b>, "SATCHMO-JS: a webserver for simultaneous protein multiple sequence alignment and phylogenetic tree construction," <i>Nucleic Acids Res</i>. 2010.
</ul>-->
<p>References cited:</p>
<ol class='refs'>

    <li><a target='_blank' href='http://www.ncbi.nlm.nih.gov/pubmed/15034147?dopt=Abstract' class='ref' name='ref1'>Edgar, R., "MUSCLE: multiple sequence alignment with high accuracy and high throughput," <i>Nucleic Acids Res</i>. 2004 Mar 19;32(5):<b>1792-7</b>.</a>
    </li>

    <li><a target='_blank' href='http://www.ncbi.nlm.nih.gov/pubmed/15152092?dopt=Abstract' class='ref' name='ref2'>Wang, G. and Dunbrack, R.L. Jr., "Scoring profile-to-profile sequence alignments," <i>Protein Sci</i>. 2004 Jun;13(6):<b>1612-26</b>.</a>
    </li>

    <li><a target='_blank' href='http://www.ncbi.nlm.nih.gov/pubmed/11847078?dopt=Abstract' class='ref' name='ref3'>
Cline, M., Hughey, R., Karplus, K. "Predicting reliable regions in protein sequence alignments," <i>Bioinformatics</i>. 2002 Feb;18(2):<b>306â€“314</b>.</a>
    </li>
</ol>

<h2 id='id_downloads'>Datasets Used in Validation Experiments And Other Downloads</h2>
<ul>
    <li>
        Download the datasets and the results of the experiments described
        in the paper (caution: files are large)
        <ul>
            <li>
                <a href="http://makana-static.berkeley.edu/app/satchmojss/supplementary_materials_for_JSS_paper.tar.gz">
                Packaged with tar and gzip (tar.gz format) [283 MB]
                </a>
            </li>
            <li>
                <a href="http://makana-static.berkeley.edu/app/satchmojss/supplementary_materials_for_JSS_paper.zip">
                Packaged as a zip file (zip format) [283 MB]
                </a>
            </li>
        </ul>
    </li> 
    <li>
        <a href="/static/apps/satchmo/download/satchmojs_scores.xls">
        Spreadsheet of SATCHMO-JS scores</a>
    </li>
    <li><a href='http://makana-static.berkeley.edu/app/satchmojss/satchmo-js_nar_figure1-1_single.png'>Figure 1.1</a></li>
    <li><a href='http://makana-static.berkeley.edu/app/satchmojss/satchmo-js_nar_figure1-2_single.png'>Figure 1.2</a></li>
    <li><a href='http://makana-static.berkeley.edu/app/satchmojss/satchmo-js_nar_figure2_double.png'>Figure 2</a></li>
</ul>
<a href='..'>Back to SATCHMO-JS Supplementary Materials Index Page</a>
{% endblock %}
